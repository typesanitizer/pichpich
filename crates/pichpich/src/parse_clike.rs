use crate::core::{Span, WithSpan};
use crate::format_utils;
use nom::Slice;
use std::fmt::{Debug, Write};
use std::sync::Arc;

pub struct FlatDocument {
    contents: Arc<String>,
    elements: Vec<FlatDocumentElement>,
}

impl FlatDocument {
    pub fn new(contents: Arc<String>) -> FlatDocument {
        let elements = FlatDocumentParser::new(contents.as_str()).run();
        FlatDocument { contents, elements }
    }
    pub fn format_snapshot(&self) -> String {
        let mut buf = String::new();
        for element in self.elements.iter() {
            let text = self.contents.as_str().slice(element.span.into_range());
            buf.write_fmt(format_args!("- {:?}: {:?}\n", element.value, text))
                .unwrap();
        }
        return buf;
    }

    pub fn format_snapshot_with_enclosing_spans(
        &self,
        magic_comments: &mut dyn Iterator<Item = &str>,
    ) -> String {
        let mut span_marks = vec![];
        for magic_comment in magic_comments {
            let (enclosing_comment_span, enclosing_group_span) =
                self.find_enclosing_spans(magic_comment);
            if let Some(enclosing_comment_span) = enclosing_comment_span {
                span_marks.push(
                    enclosing_comment_span
                        .attach_to(format!("enclosing paragraph for {}", magic_comment)),
                );
            }
            if let Some(enclosing_group_span) = enclosing_group_span {
                span_marks.push(
                    enclosing_group_span
                        .attach_to(format!("enclosing group for {}", magic_comment)),
                );
            }
        }
        return format_utils::format_snapshot_file(self.contents.as_str(), span_marks);
    }

    // The enclosing span is defined as follows.
    // - If the span is within the text of a line comment, then the returned span
    //   includes the contiguous paragraph containing the argument slice.
    // - If the span is within the text of a block comment, then the returned
    //   span is the extend of the surrounding block comment.
    //
    // Additionally, if the span is within a group (e.g. { ... }), then the span
    // for the innermost group containing the span is also returned.
    pub fn find_enclosing_spans(&self, s: &str) -> (Option<Span>, Option<Span>) {
        assert!(!s.is_empty());
        let span = Span::slice_relative(self.contents.as_str(), s);
        let element_index = self.binary_search(span.start());
        assert!(
            element_index == self.binary_search(span.end() - 1),
            "input span is unexpectedly spread out over multiple elements"
        );
        return (
            self.enclosing_comment_span(element_index),
            self.enclosing_group_span(element_index),
        );
    }

    fn enclosing_group_span(&self, central_index: usize) -> Option<Span> {
        let mut start_offset = None;
        let mut balance = 0;
        for i in (0..central_index + 1).rev() {
            match &self.elements[i].value {
                FlatDocumentElementKind::GroupStart => {
                    balance -= 1;
                    if balance == -1 {
                        start_offset = Some(self.elements[i].span.end());
                        break;
                    }
                }
                FlatDocumentElementKind::GroupEnd => {
                    balance += 1;
                }
                _ => {}
            }
        }
        if let Some(start_offset) = start_offset {
            let mut balance = 0;
            for i in central_index..self.elements.len() {
                match &self.elements[i].value {
                    FlatDocumentElementKind::GroupStart => {
                        balance -= 1;
                    }
                    FlatDocumentElementKind::GroupEnd => {
                        balance += 1;
                        if balance == 1 {
                            let end_offset = self.elements[i].span.start();
                            return Some(Span::new(start_offset, end_offset));
                        }
                    }
                    _ => {}
                }
            }
        }
        return None;
    }

    fn find_matching_comment(
        &self,
        start: usize,
        forward: bool,
        need: FlatDocumentElementKind,
        dual: FlatDocumentElementKind,
    ) -> usize {
        let mut best = start;
        let shift = (forward as isize * 2) - 1;
        let mut cursor = (start as isize).wrapping_add(shift);
        // See also NOTE(ref: regular-flat-doc-structure)
        let mut balance = 0;
        while (0..self.elements.len()).contains(&(cursor as usize)) {
            let contents = self.contents.as_str();
            match &self.elements[cursor as usize].value {
                FlatDocumentElementKind::Content => {
                    if balance == 0 {
                        let newlines = contents
                            .slice(self.elements[cursor as usize].span.into_range())
                            .as_bytes()
                            .iter()
                            .filter(|c| **c == b'\n')
                            .count();
                        if newlines > 1 {
                            break;
                        }
                    }
                }
                k if *k == need => {
                    balance += 1;
                    if balance == 0 {
                        let prev_cursor = cursor.wrapping_add(-shift);
                        let prev_content = self.elements[prev_cursor as usize];
                        assert_eq!(prev_content.value, FlatDocumentElementKind::Content);
                        let is_para_cont = contents
                            .slice(prev_content.span.into_range())
                            .contains(|c: char| !c.is_whitespace());
                        if is_para_cont {
                            best = cursor as usize;
                        } else {
                            break;
                        }
                    }
                }
                k if *k == dual => {
                    balance -= 1;
                }
                _ => break,
            }
            cursor = cursor.wrapping_add(shift);
        }
        return best;
    }

    fn enclosing_comment_span(&self, central_index: usize) -> Option<Span> {
        use FlatDocumentElementKind as Kind;
        assert!(self.elements[central_index].value == Kind::Content);
        match (
            self.elements.get(central_index - 1),
            self.elements.get(central_index + 1),
        ) {
            (Some(pre), Some(post))
                if pre.value == Kind::LineCommentStart && post.value == Kind::LineCommentEnd =>
            {
                let para_start_index = self.find_matching_comment(
                    central_index - 1,
                    false,
                    Kind::LineCommentStart,
                    Kind::LineCommentEnd,
                );
                let para_end_index = self.find_matching_comment(
                    central_index + 1,
                    true,
                    Kind::LineCommentEnd,
                    Kind::LineCommentStart,
                );
                let pre = &self.elements[para_start_index];
                let post = &self.elements[para_end_index];
                return Some(Span::new(pre.span.end(), post.span.start()));
            }
            (Some(pre), Some(post))
                if pre.value == Kind::BlockCommentStart && post.value == Kind::BlockCommentEnd =>
            {
                return Some(Span::new(pre.span.end(), post.span.start()))
            }
            (_pre, _post) => {
                return None;
            }
        }
    }
    fn binary_search(&self, offset: usize) -> usize {
        assert!(offset < self.contents.len());
        assert!(!self.elements.is_empty());
        let mut lo = 0;
        let mut hi = self.elements.len();
        while lo < hi - 1 {
            let mid = (lo + hi) / 2;
            let mid_span = self.elements[mid].span;
            if mid_span.contains(offset) {
                return mid;
            } else if offset < mid_span.start() {
                assert!(mid != lo, "offset is below minimum");
                hi = mid;
                continue;
            } else if offset >= mid_span.end() {
                assert!(mid + 1 != hi, "offset is outside maximum");
                lo = mid + 1;
            }
        }
        assert!(
            self.elements[lo].span.contains(offset),
            "offset not in middle range"
        );
        return lo;
    }
}

type FlatDocumentElement = WithSpan<FlatDocumentElementKind>;

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum FlatDocumentElementKind {
    Content,
    BlockCommentStart,
    BlockCommentEnd,
    GroupStart,
    GroupEnd,
    LineCommentStart,
    LineCommentEnd,
}

// NOTE(def: regular-flat-doc-structure): Comments always come as three elements
//   XCommentStart Content XCommentEnd
// The corresponding spans for Content and XCommentEnd may be empty,
// but using a regular pattern simplifies trying to interpret the output
// when inferring surrounding spans.

struct FlatDocumentParser<'a> {
    state: FlatDocumentParserState,
    original: &'a str,
    rest: &'a str,
    elements: Vec<FlatDocumentElement>,
}

#[allow(clippy::enum_variant_names)]
#[derive(Copy, Clone)]
enum FlatDocumentParserState {
    InBlockComment,
    InLineComment,
    NotInComment,
}

impl<'a> FlatDocumentParser<'a> {
    fn new<'b>(input: &'b str) -> FlatDocumentParser<'b> {
        FlatDocumentParser {
            state: FlatDocumentParserState::NotInComment,
            original: input,
            rest: input,
            elements: vec![],
        }
    }

    fn run(mut self) -> Vec<FlatDocumentElement> {
        while !self.rest.is_empty() {
            match self.state {
                FlatDocumentParserState::NotInComment => {
                    let mut end_index = 0;
                    let mut resume_index = 0;
                    let mut state_change = None;
                    let mut extra = None;
                    for i in 0..self.rest.len() {
                        let b = self.rest.as_bytes()[i];
                        if b == b'{' {
                            (end_index, resume_index) = (i, i + 1);
                            extra = Some(FlatDocumentElementKind::GroupStart);
                            break;
                        }
                        if b == b'}' {
                            (end_index, resume_index) = (i, i + 1);
                            extra = Some(FlatDocumentElementKind::GroupEnd);
                            break;
                        }
                        if b == b'/' {
                            if let Some(&next) = self.rest.as_bytes().get(i + 1) {
                                if next == b'*' {
                                    (end_index, resume_index) = (i, i + 2);
                                    extra = Some(FlatDocumentElementKind::BlockCommentStart);
                                    state_change = Some(FlatDocumentParserState::InBlockComment);
                                    break;
                                }
                                if next == b'/' {
                                    (end_index, resume_index) = (i, i + 2);
                                    extra = Some(FlatDocumentElementKind::LineCommentStart);
                                    state_change = Some(FlatDocumentParserState::InLineComment);
                                    break;
                                }
                            }
                        }
                        (end_index, resume_index) = (i, i + 1);
                    }
                    let content =
                        Span::slice_relative(self.original, self.rest.slice(0..end_index))
                            .attach_to(FlatDocumentElementKind::Content);
                    self.elements.push(content);
                    if let Some(next_element) = extra {
                        let element = Span::slice_relative(
                            self.original,
                            self.rest.slice(end_index..resume_index),
                        )
                        .attach_to(next_element);
                        self.elements.push(element);
                    }
                    self.rest = self.rest.slice(resume_index..);
                    if let Some(state) = state_change {
                        self.state = state;
                    }
                }
                FlatDocumentParserState::InBlockComment => {
                    let (end_index, resume_index) = {
                        if let Some(end_index) = self.rest.find("*/") {
                            (end_index, end_index + 2)
                        } else {
                            (self.rest.len(), self.rest.len())
                        }
                    };
                    // See also NOTE(ref: regular-flat-doc-structure)
                    let content =
                        Span::slice_relative(self.original, self.rest.slice(0..end_index))
                            .attach_to(FlatDocumentElementKind::Content);
                    let end = Span::slice_relative(
                        self.original,
                        self.rest.slice(end_index..resume_index),
                    )
                    .attach_to(FlatDocumentElementKind::BlockCommentEnd);
                    self.elements.push(content);
                    self.elements.push(end);
                    self.state = FlatDocumentParserState::NotInComment;
                    self.rest = self.rest.slice(resume_index..);
                }
                FlatDocumentParserState::InLineComment => {
                    let end_index = self.rest.find('\n').unwrap_or(self.rest.len());
                    // See also NOTE(ref: regular-flat-doc-structure)
                    let content =
                        Span::slice_relative(self.original, self.rest.slice(0..end_index))
                            .attach_to(FlatDocumentElementKind::Content);
                    let end =
                        Span::slice_relative(self.original, self.rest.slice(end_index..end_index))
                            .attach_to(FlatDocumentElementKind::LineCommentEnd);
                    self.elements.push(content);
                    self.elements.push(end);
                    self.state = FlatDocumentParserState::NotInComment;
                    self.rest = self.rest.slice(end_index..);
                }
            }
        }
        return self.elements;
    }
}
