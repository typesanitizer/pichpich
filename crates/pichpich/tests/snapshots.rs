// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

use miette::{GraphicalReportHandler, GraphicalTheme};
use pichpich::frontend::Options;
use pichpich::parse_clike::FlatDocument;
use pichpich::{frontend, main_impl, parse_clike};
use pichpich_config::ErrorConfig;
use std::path::PathBuf;
use std::sync::Arc;

fn for_each_path(pattern: &str, test_path: &dyn Fn(PathBuf) -> ()) {
    for entry in glob::glob(pattern).expect("failed to read pattern") {
        let path = match entry {
            Ok(p) => p,
            Err(e) => {
                panic!("ill-formed result from glob {}", e);
            }
        };
        test_path(path)
    }
}

#[test]
fn enclosing_spans_snapshots() {
    let opts = Options {
        root: "tests/snapshots/enclosing-spans".parse().unwrap(),
        error_config: ErrorConfig::default(),
    };
    let syntax_data = frontend::gather(&opts).0;
    assert!(!syntax_data.path_to_comment_map.is_empty());
    let mut path_list: Vec<_> = syntax_data
        .path_to_content_map
        .iter()
        .map(|(p, _)| p.clone())
        .collect();
    path_list.sort();
    assert!(path_list.len() == 2);
    let mut documents = vec![];
    for path in path_list.iter() {
        let contents = syntax_data.path_to_content_map.get(path).unwrap().clone();
        documents.push(FlatDocument::new(contents));
    }
    for (i, path) in path_list.iter().enumerate() {
        let contents = syntax_data.path_to_content_map.get(path).unwrap();
        let comments = syntax_data.path_to_comment_map.get(path).unwrap();
        let mut comment_texts = comments.iter().map(|comment| {
            contents
                .as_str()
                .get(comment.span.into_range())
                .expect("stored out-of-bounds indexes in comment span")
        });
        insta::assert_snapshot!(
            documents[i].format_snapshot_with_enclosing_spans(&mut comment_texts)
        );
    }
}

#[test]
fn doc_parsing_snapshots() {
    for_each_path("tests/snapshots/parse-*", &|path| {
        let file_contents = std::fs::read_to_string(&path).unwrap();
        let document = parse_clike::FlatDocument::new(Arc::new(file_contents));
        insta::assert_snapshot!(document.format_snapshot())
    })
}

#[test]
fn internal_state_snapshots() {
    for_each_path("tests/snapshots/internal-state-*", &|path| {
        let opts = Options {
            root: path.to_owned(),
            error_config: ErrorConfig::default(),
        };
        let result = frontend::gather(&opts);
        insta::assert_snapshot!(result.0.format_snapshot());
    });
}

#[test]
fn error_snapshots() {
    // Avoid colors in output to make snapshot files readable in a standalone way.
    let reporter = GraphicalReportHandler::new_themed(GraphicalTheme::unicode_nocolor());
    for_each_path("tests/snapshots/error-*", &|path| {
        let mut error_config = ErrorConfig::default();
        assert!(error_config.populate(vec!["error:all".to_string()]).is_ok());
        let opts = Options {
            root: path.to_owned(),
            error_config,
        };
        let result = main_impl(opts);
        match result {
            Ok(()) => panic!(
                "expected errors when processing {:?}",
                path.to_string_lossy()
            ),
            Err(e) => {
                let mut buf = String::new();
                assert!(
                    reporter.render_report(&mut buf, e.as_ref()).is_ok(),
                    "rendering failed"
                );
                insta::assert_snapshot!(buf);
            }
        }
    });
}
