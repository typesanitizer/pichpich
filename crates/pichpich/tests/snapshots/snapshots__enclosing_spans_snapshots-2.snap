---
source: crates/pichpich/tests/snapshots.rs
expression: "documents[i].format_snapshot_with_enclosing_spans(&mut comment_texts)"
---

// Matrix of test cases for line comments:
// - [top-level, in braces]
// - [paragraph start, NOT paragraph start]
// - [paragraph end, NOT paragraph end]
// - [line start, NOT line start]
//
// So it is a total of 16 cases.

//-----------------------------------------------------------------------------
// ( 1) top-level, paragraph start, paragraph end, line start

// NOTE(def: lc-1-1) aa
  ^^^^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-1-1)

//
// NOTE(def: lc-1-2) bb
  ^^^^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-1-2)

// NOTE(def: lc-1-3)
  ^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-1-3)
//

//
// NOTE(def: lc-1-4)
  ^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-1-4)
//

//-----------------------------------------------------------------------------
// ( 2) top-level, paragraph start, paragraph end, NOT line start

// This is NOTE(def: lc-2-1) aa
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-2-1)

//
// This is NOTE(def: lc-2-2) bb
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-2-2)

// This is NOTE(def: lc-2-3)
  ^^^^^^^^^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-2-3)
//

//
// This is NOTE(def: lc-2-4)
  ^^^^^^^^^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-2-4)
//

//-----------------------------------------------------------------------------
// ( 3) top-level, paragraph start, NOT paragraph end, line start

// NOTE(def: lc-3-1) aa
  ^ start enclosing paragraph for NOTE(def: lc-3-1)
// Blah blah
            ^ end enclosing paragraph for NOTE(def: lc-3-1)

//
// NOTE(def: lc-3-2) bb
  ^ start enclosing paragraph for NOTE(def: lc-3-2)
// Blah blah
            ^ end enclosing paragraph for NOTE(def: lc-3-2)

// NOTE(def: lc-3-3)
  ^ start enclosing paragraph for NOTE(def: lc-3-3)
// Blah blah
            ^ end enclosing paragraph for NOTE(def: lc-3-3)
//

//
// NOTE(def: lc-3-4)
  ^ start enclosing paragraph for NOTE(def: lc-3-4)
// Blah blah
            ^ end enclosing paragraph for NOTE(def: lc-3-4)
//

//-----------------------------------------------------------------------------
// ( 4) top-level, paragraph start, NOT paragraph end, NOT line start

// This is NOTE(def: lc-4-1) aa
  ^ start enclosing paragraph for NOTE(def: lc-4-1)
// Blah blah
            ^ end enclosing paragraph for NOTE(def: lc-4-1)

//
// This is NOTE(def: lc-4-2) bb
  ^ start enclosing paragraph for NOTE(def: lc-4-2)
// Blah blah
            ^ end enclosing paragraph for NOTE(def: lc-4-2)

// This is NOTE(def: lc-4-3)
  ^ start enclosing paragraph for NOTE(def: lc-4-3)
// Blah blah
            ^ end enclosing paragraph for NOTE(def: lc-4-3)
//

//
// This is NOTE(def: lc-4-4)
  ^ start enclosing paragraph for NOTE(def: lc-4-4)
// Blah blah
            ^ end enclosing paragraph for NOTE(def: lc-4-4)
//

//-----------------------------------------------------------------------------
// ( 5) top-level, NOT paragraph start, paragraph end, line start

// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-5-1)
// NOTE(def: lc-5-1) aa
                       ^ end enclosing paragraph for NOTE(def: lc-5-1)

//
// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-5-2)
// NOTE(def: lc-5-2) bb
                       ^ end enclosing paragraph for NOTE(def: lc-5-2)

// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-5-3)
// NOTE(def: lc-5-3)
                    ^ end enclosing paragraph for NOTE(def: lc-5-3)
//

//
// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-5-4)
// NOTE(def: lc-5-4)
                    ^ end enclosing paragraph for NOTE(def: lc-5-4)
//

//-----------------------------------------------------------------------------
// ( 6) top-level, NOT paragraph start, paragraph end, NOT line start

// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-6-1)
// This is NOTE(def: lc-6-1) aa
                               ^ end enclosing paragraph for NOTE(def: lc-6-1)

//
// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-6-2)
// This is NOTE(def: lc-6-2) bb
                               ^ end enclosing paragraph for NOTE(def: lc-6-2)

// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-6-3)
// This is NOTE(def: lc-6-3)
                            ^ end enclosing paragraph for NOTE(def: lc-6-3)
//

//
// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-6-4)
// This is NOTE(def: lc-6-4)
                            ^ end enclosing paragraph for NOTE(def: lc-6-4)
//

//-----------------------------------------------------------------------------
// ( 7) top-level, NOT paragraph start, NOT paragraph end, line start

// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-7-1)
// NOTE(def: lc-7-1) aa
// Even more blah blah
                      ^ end enclosing paragraph for NOTE(def: lc-7-1)

//
// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-7-2)
// NOTE(def: lc-7-2) bb
// Even more blah blah
                      ^ end enclosing paragraph for NOTE(def: lc-7-2)

// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-7-3)
// NOTE(def: lc-7-3)
// Even more blah blah
                      ^ end enclosing paragraph for NOTE(def: lc-7-3)
//

//
// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-7-4)
// NOTE(def: lc-7-4)
// Even more blah blah
                      ^ end enclosing paragraph for NOTE(def: lc-7-4)
//

//-----------------------------------------------------------------------------
// ( 8) top-level, NOT paragraph start, NOT paragraph end, NOT line start

// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-8-1)
// This is NOTE(def: lc-8-1) aa
// Even more blah blah
                      ^ end enclosing paragraph for NOTE(def: lc-8-1)

//
// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-8-2)
// This is NOTE(def: lc-8-2) bb
// Even more blah blah
                      ^ end enclosing paragraph for NOTE(def: lc-8-2)

// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-8-3)
// This is NOTE(def: lc-8-3)
// Even more blah blah
                      ^ end enclosing paragraph for NOTE(def: lc-8-3)
//

//
// Blah blah
  ^ start enclosing paragraph for NOTE(def: lc-8-4)
// This is NOTE(def: lc-8-4)
// Even more blah blah
                      ^ end enclosing paragraph for NOTE(def: lc-8-4)
//

//-----------------------------------------------------------------------------
// ( 9) in group, paragraph start, paragraph end, line start

{
  {
    // NOTE(def: lc-9-1) aa
< start enclosing group for NOTE(def: lc-9-1)
< start enclosing group for NOTE(def: lc-9-2)
< start enclosing group for NOTE(def: lc-9-3)
< start enclosing group for NOTE(def: lc-9-4)
      ^^^^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-9-1)
    { }
    //
    // NOTE(def: lc-9-2) bb
      ^^^^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-9-2)

    // NOTE(def: lc-9-3)
      ^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-9-3)
    //
    { {} }
    //
    // NOTE(def: lc-9-4)
      ^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-9-4)
    //
  }
  ^ end enclosing group for NOTE(def: lc-9-1)
  ^ end enclosing group for NOTE(def: lc-9-2)
  ^ end enclosing group for NOTE(def: lc-9-3)
  ^ end enclosing group for NOTE(def: lc-9-4)
}

//-----------------------------------------------------------------------------
// (10) in group, paragraph start, paragraph end, NOT line start

{
  {
    // This is NOTE(def: lc-10-1) aa
< start enclosing group for NOTE(def: lc-10-1)
< start enclosing group for NOTE(def: lc-10-2)
< start enclosing group for NOTE(def: lc-10-3)
< start enclosing group for NOTE(def: lc-10-4)
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-10-1)
    { }
    //
    // This is NOTE(def: lc-10-2) bb
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-10-2)

    // This is NOTE(def: lc-10-3)
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-10-3)
    //
    { {} }
    //
    // This is NOTE(def: lc-10-4)
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^ enclosing paragraph for NOTE(def: lc-10-4)
    //
  }
  ^ end enclosing group for NOTE(def: lc-10-1)
  ^ end enclosing group for NOTE(def: lc-10-2)
  ^ end enclosing group for NOTE(def: lc-10-3)
  ^ end enclosing group for NOTE(def: lc-10-4)
}

//-----------------------------------------------------------------------------
// (11) in group, paragraph start, NOT paragraph end, line start

{
  {
    // NOTE(def: lc-11-1) aa
< start enclosing group for NOTE(def: lc-11-1)
< start enclosing group for NOTE(def: lc-11-2)
< start enclosing group for NOTE(def: lc-11-3)
< start enclosing group for NOTE(def: lc-11-4)
      ^ start enclosing paragraph for NOTE(def: lc-11-1)
    // Blah blah
                ^ end enclosing paragraph for NOTE(def: lc-11-1)
    { }
    //
    // NOTE(def: lc-11-2) bb
      ^ start enclosing paragraph for NOTE(def: lc-11-2)
    // Blah blah
                ^ end enclosing paragraph for NOTE(def: lc-11-2)

    // NOTE(def: lc-11-3)
      ^ start enclosing paragraph for NOTE(def: lc-11-3)
    // Blah blah
                ^ end enclosing paragraph for NOTE(def: lc-11-3)
    //
    { {} }
    //
    // NOTE(def: lc-11-4)
      ^ start enclosing paragraph for NOTE(def: lc-11-4)
    // Blah blah
                ^ end enclosing paragraph for NOTE(def: lc-11-4)
    //
  }
  ^ end enclosing group for NOTE(def: lc-11-1)
  ^ end enclosing group for NOTE(def: lc-11-2)
  ^ end enclosing group for NOTE(def: lc-11-3)
  ^ end enclosing group for NOTE(def: lc-11-4)
}

//-----------------------------------------------------------------------------
// (12) in group, paragraph start, NOT paragraph end, NOT line start

{
  {
    // This is NOTE(def: lc-12-1) aa
< start enclosing group for NOTE(def: lc-12-1)
< start enclosing group for NOTE(def: lc-12-2)
< start enclosing group for NOTE(def: lc-12-3)
< start enclosing group for NOTE(def: lc-12-4)
      ^ start enclosing paragraph for NOTE(def: lc-12-1)
    // Blah blah
                ^ end enclosing paragraph for NOTE(def: lc-12-1)
    { }
    //
    // This is NOTE(def: lc-12-2) bb
      ^ start enclosing paragraph for NOTE(def: lc-12-2)
    // Blah blah
                ^ end enclosing paragraph for NOTE(def: lc-12-2)

    // This is NOTE(def: lc-12-3)
      ^ start enclosing paragraph for NOTE(def: lc-12-3)
    // Blah blah
                ^ end enclosing paragraph for NOTE(def: lc-12-3)
    //
    { {} }
    //
    // This is NOTE(def: lc-12-4)
      ^ start enclosing paragraph for NOTE(def: lc-12-4)
    // Blah blah
                ^ end enclosing paragraph for NOTE(def: lc-12-4)
    //
  }
  ^ end enclosing group for NOTE(def: lc-12-1)
  ^ end enclosing group for NOTE(def: lc-12-2)
  ^ end enclosing group for NOTE(def: lc-12-3)
  ^ end enclosing group for NOTE(def: lc-12-4)
}


//-----------------------------------------------------------------------------
// (13) in group, NOT paragraph start, paragraph end, line start

{
  {
    // Blah blah
< start enclosing group for NOTE(def: lc-13-1)
< start enclosing group for NOTE(def: lc-13-2)
< start enclosing group for NOTE(def: lc-13-3)
< start enclosing group for NOTE(def: lc-13-4)
      ^ start enclosing paragraph for NOTE(def: lc-13-1)
    // NOTE(def: lc-13-1) aa
                            ^ end enclosing paragraph for NOTE(def: lc-13-1)
    { }
    //
    // Blah blah
      ^ start enclosing paragraph for NOTE(def: lc-13-2)
    // NOTE(def: lc-13-2) bb
                            ^ end enclosing paragraph for NOTE(def: lc-13-2)

    // Blah blah
      ^ start enclosing paragraph for NOTE(def: lc-13-3)
    // NOTE(def: lc-13-3)
                         ^ end enclosing paragraph for NOTE(def: lc-13-3)
    //
    { {} }
    //
    // Blah blah
      ^ start enclosing paragraph for NOTE(def: lc-13-4)
    // NOTE(def: lc-13-4)
                         ^ end enclosing paragraph for NOTE(def: lc-13-4)
    //
  }
  ^ end enclosing group for NOTE(def: lc-13-1)
  ^ end enclosing group for NOTE(def: lc-13-2)
  ^ end enclosing group for NOTE(def: lc-13-3)
  ^ end enclosing group for NOTE(def: lc-13-4)
}

//-----------------------------------------------------------------------------
// (14) in group, NOT paragraph start, paragraph end, NOT line start

{
  {
    // Blah blah
< start enclosing group for NOTE(def: lc-14-1)
< start enclosing group for NOTE(def: lc-14-2)
< start enclosing group for NOTE(def: lc-14-3)
< start enclosing group for NOTE(def: lc-14-4)
      ^ start enclosing paragraph for NOTE(def: lc-14-1)
    // This is NOTE(def: lc-14-1) aa
                                    ^ end enclosing paragraph for NOTE(def: lc-14-1)
    { }
    //
    // Blah blah
      ^ start enclosing paragraph for NOTE(def: lc-14-2)
    // This is NOTE(def: lc-14-2) bb
                                    ^ end enclosing paragraph for NOTE(def: lc-14-2)

    // Blah blah
      ^ start enclosing paragraph for NOTE(def: lc-14-3)
    // This is NOTE(def: lc-14-3)
                                 ^ end enclosing paragraph for NOTE(def: lc-14-3)
    //
    { {} }
    //
    // Blah blah
      ^ start enclosing paragraph for NOTE(def: lc-14-4)
    // This is NOTE(def: lc-14-4)
                                 ^ end enclosing paragraph for NOTE(def: lc-14-4)
    //
  }
  ^ end enclosing group for NOTE(def: lc-14-1)
  ^ end enclosing group for NOTE(def: lc-14-2)
  ^ end enclosing group for NOTE(def: lc-14-3)
  ^ end enclosing group for NOTE(def: lc-14-4)
}

//-----------------------------------------------------------------------------
// (15) in group, NOT paragraph start, NOT paragraph end, line start

{
  {
    // Blah blah
< start enclosing group for NOTE(def: lc-15-1)
< start enclosing group for NOTE(def: lc-15-2)
< start enclosing group for NOTE(def: lc-15-3)
< start enclosing group for NOTE(def: lc-15-4)
      ^ start enclosing paragraph for NOTE(def: lc-15-1)
    // NOTE(def: lc-15-1) aa
    // Even more blah blah
                          ^ end enclosing paragraph for NOTE(def: lc-15-1)
    { }
    //
    // Blah blah
      ^ start enclosing paragraph for NOTE(def: lc-15-2)
    // NOTE(def: lc-15-2) bb
    // Even more blah blah
                          ^ end enclosing paragraph for NOTE(def: lc-15-2)

    // Blah blah
      ^ start enclosing paragraph for NOTE(def: lc-15-3)
    // NOTE(def: lc-15-3)
    // Even more blah blah
                          ^ end enclosing paragraph for NOTE(def: lc-15-3)
    //
    { {} }
    //
    // Blah blah
      ^ start enclosing paragraph for NOTE(def: lc-15-4)
    // NOTE(def: lc-15-4)
    // Even more blah blah
                          ^ end enclosing paragraph for NOTE(def: lc-15-4)
    //
  }
  ^ end enclosing group for NOTE(def: lc-15-1)
  ^ end enclosing group for NOTE(def: lc-15-2)
  ^ end enclosing group for NOTE(def: lc-15-3)
  ^ end enclosing group for NOTE(def: lc-15-4)
}

//-----------------------------------------------------------------------------
// (16) in group, NOT paragraph start, NOT paragraph end, NOT line start

{
  {
    // Blah blah
< start enclosing group for NOTE(def: lc-8-1)
< start enclosing group for NOTE(def: lc-8-2)
< start enclosing group for NOTE(def: lc-8-3)
< start enclosing group for NOTE(def: lc-8-4)
      ^ start enclosing paragraph for NOTE(def: lc-8-1)
    // This is NOTE(def: lc-8-1) aa
    // Even more blah blah
                          ^ end enclosing paragraph for NOTE(def: lc-8-1)
    { }
    //
    // Blah blah
      ^ start enclosing paragraph for NOTE(def: lc-8-2)
    // This is NOTE(def: lc-8-2) bb
    // Even more blah blah
                          ^ end enclosing paragraph for NOTE(def: lc-8-2)

    // Blah blah
      ^ start enclosing paragraph for NOTE(def: lc-8-3)
    // This is NOTE(def: lc-8-3)
    // Even more blah blah
                          ^ end enclosing paragraph for NOTE(def: lc-8-3)
    //
    { {} }
    //
    // Blah blah
      ^ start enclosing paragraph for NOTE(def: lc-8-4)
    // This is NOTE(def: lc-8-4)
    // Even more blah blah
                          ^ end enclosing paragraph for NOTE(def: lc-8-4)
    //
  }
  ^ end enclosing group for NOTE(def: lc-8-1)
  ^ end enclosing group for NOTE(def: lc-8-2)
  ^ end enclosing group for NOTE(def: lc-8-3)
  ^ end enclosing group for NOTE(def: lc-8-4)
}

